src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DecisionDagRewriter.cs

```diff
@@ -410,53 +412,510 @@ private void GenerateTest(BoundExpression test, BoundDecisionDagNode whenTrue, B
-                if (firstTest.Input.Type.SpecialType == SpecialType.System_String)
+                /// <summary>
+                /// A node representing a final destination that requires no further dispatch.
+                /// </summary>
+                internal class LeafDispatchNode : ValueDispatchNode
                 {
-                    EnsureStringHashFunction(cases.Count, node.Syntax);
-                    stringEquality = _localRewriter.UnsafeGetSpecialTypeMethod(node.Syntax, SpecialMember.System_String__op_Equality);
+                    public readonly LabelSymbol Label;
+                    public LeafDispatchNode(SyntaxNode syntax, LabelSymbol Label) : base(syntax) => this.Label = Label;
+                    public override string ToString() => "Leaf";
                 }
 
-                LabelSymbol defaultLabel = GetDagNodeLabel(previous.WhenFalse);
-                var dispatch = new BoundSwitchDispatch(
-                    node.Syntax, _tempAllocator.GetTemp(firstTest.Input), cases.ToImmutableAndFree(), defaultLabel, stringEquality);
-                _loweredDecisionDag.Add(dispatch);
-                return true;
+                /// <summary>
+                /// A node representing a dispatch based on a relational test of the input value by some constant.
+                /// Nodes of this kind are required to be height-balanced when constructed, so that when the full
+                /// decision tree is produced it generates a balanced tree of comparisons.
+                /// </summary>
+                internal class RelationalDispatch : ValueDispatchNode
+                {
+                    private int _height;
+                    protected override int Height => _height;
+                    public readonly ConstantValue Value;
+                    public readonly BinaryOperatorKind Operator;
+                    /// <summary>The side of the test handling lower values. The true side for &lt; and &lt;=, the false side for > and >=.</summary>
+                    private ValueDispatchNode Left { get; set; }
+                    /// <summary>The side of the test handling higher values. The false side for &lt; and &lt;=, the true side for > and >=.</summary>
+                    private ValueDispatchNode Right { get; set; }
+                    private RelationalDispatch(SyntaxNode syntax, ConstantValue value, BinaryOperatorKind op, ValueDispatchNode left, ValueDispatchNode right) : base(syntax)
+                    {
+                        this.Value = value;
+                        this.Operator = op;
+                        WithLeftAndRight(left, right);
+                    }
+                    public ValueDispatchNode WhenTrue => IsReversed(Operator) ? Right : Left;
+                    public ValueDispatchNode WhenFalse => IsReversed(Operator) ? Left : Right;
+                    public override string ToString() => $"RelationalDispatch.{Height}({Left} {Operator.Operator()} {Value} {Right})";
```

Comment:
>Operator.Operator() [](start = 95, length = 19)

There are a number of calls to `.Operator()` in this type, sometimes redundantly (such as some callers of `IsReversed()` and within `IsReversed()`). Consider asserting the `BinaryOperatorKind` is just the `Operator()` part and removing all the calls.

Reply:
It isn't just the operator part. Almost everywhere in this class it is the operator plus the type code. I will add some assertions to clarify that.

---
In reply to: [391910738](https://github.com/dotnet/roslyn/pull/42313#discussion_r391910738) [](ancestors = 391910738)